File: ruby/GameUniverse.rb
#encoding-utf:8

require_relative 'Dice'
require_relative 'GameStateController'
require_relative 'CardDealer'
require_relative 'SpaceStation'
require_relative 'GameUniverseToUI'
require_relative 'GameCharacter'
require_relative 'CombatResult'

module Deepspace
    class GameUniverse
        @@WIN = 10
        def initialize
            @turns = 0
            @currentStationIndex = 0
            @dice = Dice.new
            @gameState = GameStateController.new
            @currentStation = nil
            @currentEnemy = nil
            @spaceStations  = Array.new
        end

        def combatGo(station, enemy)
            ch = @dice.firstShot
            if (ch == GameCharacter::ENEMYSTARSHIP)
                fire = enemy.fire
                result = station.receiveShot(fire)
                if (result == ShotResult::RESIST)
                    fire = station.fire
                    result = enemy.receiveShot(fire)
                    enemyWins = (result == ShotResult::RESIST)
                else
                    enemyWins = true
                end
            else
                fire = station.fire
                result = enemy.receiveShot(fire)
                enemyWins = (result == ShotResult::RESIST)
            end

            if enemyWins
                s = station.speed
                moves = @dice.spaceStationMoves(s)
                if !moves
                    damage = enemy.damage
                    station.setPendingDamage(damage)
                    combatResult = CombatResult::ENEMYWINS
                else
                    station.move
                    combatResult = CombatResult::STATIONESCAPES
                end
            else
                aLoot = enemy.loot
                station.setLoot(aLoot)
                combatResult = CombatResult::STATIONWINS
            end
        
            @gameState.next(@turns,@spaceStations.length)
        
            return combatResult

        end

        public
        def combat
            state = @gameState.state
            if (state == GameState::BEFORECOMBAT || state == GameState::INIT)
                return combatGo(@currentStation,@currentEnemy)
            else
                return CombatResult::NOCOMBAT
            end
        end 

        def discardHangar
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.discardHangar
            end
        end

        def discardShieldBooster(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.discardShieldBooster(i)
            end
        end

        def discardShieldBoosterInHangar(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.discardShieldBoosterInHangar(i)
            end
        end

        def discardWeapon(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.discardWeapon(i)
            end
        end

        def discardWeaponInHangar(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.discardWeaponInHangar(i)
            end
        end

        def state
            return @gameState.state       #devuelve gameState o gameState.state?: segun el diagrama es lo segundo
        end

        def getUIversion
            return GameUniverseToUI.new(@currentStation, @currentEnemy)
        end

        def haveAWinner
            return @currentStation.nMedals >= @@WIN
        end

        def init(names) #array de strings
            state = @gameState.state
            size = names.length

            if state == Deepspace::GameState::CANNOTPLAY
                dealer = CardDealer.instance

                for i in (0..size-1)
                    supplies = dealer.nextSuppliesPackage
                    station = SpaceStation.new(names[i], supplies)

                    nh = @dice.initWithNHangars
                    nw = @dice.initWithNWeapons
                    ns = @dice.initWithNShields

                    lo = Loot.new(0, nw, ns, nh, 0)
                    station.setLoot(lo)
                
                    @spaceStations.push(station)
                end

                @currentStationIndex = @dice.whoStarts(size)
                @currentStation = @spaceStations[@currentStationIndex]
                @currentEnemy = dealer.nextEnemy
                @gameState.next(@turns,size)
            end
        end

        def mountShieldBooster(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.mountShieldBooster(i)
            end
        end

        def mountWeapon(i)
            if(state == GameState::INIT || state == GameState::AFTERCOMBAT)
                @currentStation.mountWeapon(i)
            end
        end

        def nextTurn
            state = @gameState.state

            if state == Deepspace::GameState::AFTERCOMBAT
                stationState = @currentStation.validState()
                if stationState
                    @currentStationIndex = (@currentStationIndex+1) % @spaceStations.length
                    @turns+=1
                    @currentStation=@spaceStations[@currentStationIndex]
                    @currentStation.cleanUpMountedItems

                    dealer=CardDealer.instance
                    @currentEnemy = dealer.nextEnemy
                    @gameState.next(@turns, @spaceStations.length)

                    return true
                end
                return false
            end
            return false
        end

        def to_s
            if state == Deepspace::GameState::INIT
              state_ = "state:=INIT\n"
            elsif state == Deepspace::GameState::CANNOTPLAY
              state_ = "state:=CANNOTPLAT\n"
            elsif state == Deepspace::GameState::BEFORECOMBAT
              state_ = "state:=BEFORECOMBAT\n"
            else
              state_ = "state:=AFTERCOMBAT\n"
            end
      
            actualEnemy = "CURRENTENEMY:=nil\n"
            if @currentEnemy != nil
              actualEnemy = "CURRENTENEMY:\n << #{@currentEnemy.to_s}"
            end
      
            dice_ = "DICE:\n #{@dice.to_s}"
      
            actualStation = "CURRENTSTATION:=nil\n"
            if @currentStation != nil
              actualStation = "CURRENTSTATION:\n #{@currentStation.to_s}"
            end
      
            stations = "SPACESTATIONS: \n #{@spaceStations.to_s.inspect}"
      
            return "INDEX: " + "#{@currentStationIndex}" + "\n"+ "TURNS: " + "#{@turns}" + "\n" + state_ + actualEnemy + actualStation + dice_ + stations + "\n"
          end
    end
end

File: java/Deepspace/src/deepspace/GameUniverse.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;
import java.util.ArrayList;
/**
 *
 * @author alvaro
 */
public class GameUniverse {
    private static final int WIN = 10; 
    private int currentStationIndex;
    private int turns;
    private SpaceStation currentStation;
    private ArrayList<SpaceStation> spaceStations = new ArrayList<>();
    private GameStateController gameState;
    private EnemyStarShip currentEnemy;
    private Dice dice;
    
    public GameUniverse(){
        gameState = new GameStateController();
        turns = 0;
        dice = new Dice();
        
        currentStationIndex = 0;
        currentStation=null;
        currentEnemy=null;
    }
    
    private CombatResult combat(SpaceStation station, EnemyStarShip enemy){
        GameCharacter ch = dice.firstShot();
        boolean enemyWins, moves;
        float fire, s;
        ShotResult result;
        Damage damage;
        CombatResult combatResult;
                
        if( ch == GameCharacter.ENEMYSTARSHIP){
            fire = enemy.fire();
            result = station.receiveShot(fire);
            if (result == ShotResult.RESIST){
                fire = station.fire();
                result = enemy.receiveShot(fire);
    
                enemyWins = (result == ShotResult.RESIST);
            }else{
                enemyWins = true;
            }
        }else{
            fire = station.fire();
            result = enemy.receiveShot(fire);
            enemyWins = (result == ShotResult.RESIST);
        }
        
        if (enemyWins){
            s = station.getSpeed();
            moves = dice.spaceStationMoves(s);
            if(!moves){
                damage = enemy.getDamage();
                station.setPendingDamage(damage);
                combatResult = CombatResult.ENEMYWINS;
            }else{
                station.move();
                combatResult = CombatResult.STATIONESCAPES;
            }
        }else{
            Loot aloot = enemy.getLoot();
            station.setLoot(aloot);
            combatResult = CombatResult.STATIONWINS;
        }
        
        gameState.next(turns, spaceStations.size());
        
        return combatResult;
    }
    
    public CombatResult combat(){
        GameState state = gameState.getState();
        if( (state==GameState.BEFORECOMBAT) || (state==GameState.INIT) )
            return combat(currentStation, currentEnemy);
        else
            return CombatResult.NOCOMBAT;
    }
    
    public void discardHangar(){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.discardHangar();
    }
    
    public void discardShieldBooster(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBooster(i);
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBoosterInHangar(i);
    }
    
    public void discardWeapon(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeapon(i);
    }
    
    public void discardWeaponInHangar(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeaponInHangar(i);
    }
    
    public GameState getState(){
         return gameState.getState();
    }
    
    public GameUniverseToUI getUIversion(){
        return new GameUniverseToUI(currentStation, currentEnemy);
    }
    
    public boolean haveAWinner(){
        if(currentStation.getNMedals() >= WIN)
            return true;
        else
            return false;
    }
    
    public void init(ArrayList<String> names){
        GameState state = getState();
        
         if(state==GameState.CANNOTPLAY){
            CardDealer dealer = CardDealer.getInstance();
             
            for(int i=0; i<names.size(); i++){
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(names.get(i), supplies);
            
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                Loot l = new Loot(0,nw,ns,nh,0);
                
                station.setLoot(l);
                
                spaceStations.add(station);
            }
             
            currentStationIndex = dice.whoStarts(names.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            
            gameState.next(turns, spaceStations.size());
         }
    }
    
    public void mountShieldBooster(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.mountShieldBooster(i);
    }
    
    public void mountWeapon(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT)
            currentStation.mountWeapon(i);
    }   
    
    public boolean nextTurn(){
        GameState state = gameState.getState();
        
        if ( state == GameState.AFTERCOMBAT ){
            boolean stationState = currentStation.validState();
            if(stationState){
                currentStationIndex = (currentStationIndex+1)%spaceStations.size();
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                
                gameState.next(turns, spaceStations.size());
                
                return true;
            } else return false;
        } else return false;
    }
    
    @Override
    public String toString(){
        String actualStationInd = "Station Index:" + currentStationIndex + "\n";
        String turn = "Turns: " + turns + "\n";
        String gState = "GameState: " + gameState + "\n";
        String actualEnemy = "CurrentEnemy: null\n";
        if(currentEnemy != null){
            actualEnemy = "CurrentEnemy: \n" + currentEnemy.toString() + "\n\n";
        }
        String actualStation = "CurrentStation: null\n";
        if(currentStation != null){
            actualStation = "CurrentStation: \n" + currentStation.toString() + "\n\n";
        }
        //String dice = "Dice:  \n" + dice.toString() + "\n\n";
        
        String spaceStats = "SpaceStations: \n" + spaceStations.toString() + "\n\n";
        
        String result = actualStationInd + turn + gState + actualEnemy + actualStation  + spaceStats;
        
        return result;  
    }
    
}
